src/gui_common/ModalManager.cs

```diff
@@ -125,6 +132,28 @@ public bool HideTopMostPopup()
         return null;
     }
 
+    /// <summary>
+    ///   Attempts to hide the given modal.
+    ///   This won't hide the modal if it's exclusive and doesn't allow closing on escape.
+    /// </summary>
+    /// <param name="popup">Modal to hide</param>
+    /// <returns>True if the modal was hidden</returns>
+    private static bool HideModal(TopLevelContainer popup)
+    {
+        if (popup.Exclusive && !popup.ExclusiveAllowCloseOnEscape)
+            return false;
+
+        // This is emitted before closing to allow window using components to differentiate between "cancel" and
+        // "any other reason for closing" in case some logic can be simplified by handling just those two situations.
+        if (popup is CustomWindow dialog)
+            dialog.EmitSignal(nameof(CustomWindow.Cancelled));
+
+        popup.Close();
+        popup.Notification(Control.NotificationModalClose);
```

Comment:
Also, we want to make sure this method correctly handle already already-hidden popups. Otherwise, it would potentially cause unexpected behaviors.

Reply:
Checking popup visibility now